pennylane/gradients/parameter_shift.py

```diff
@@ -140,34 +140,72 @@ def _extract_unshifted(recipe, at_least_one_unshifted, f0, gradient_tapes, tape)
     return recipe, at_least_one_unshifted, unshifted_coeff
 
 
-def _evaluate_gradient(res, data, broadcast, r0, scalar_qfunc_output):
+def _unshifted_coeff(g, unshifted_coeff, r0):
+    """Auxiliary function; if unshifted term exists, add its contribution."""
+    if unshifted_coeff is not None:
+        # add the unshifted term
+        g = g + unshifted_coeff * r0
+    return g
+
+
+def _evaluate_gradient(tape, res, data, broadcast, r0, scalar_qfunc_output):
     """Use shifted tape evaluations and parameter-shift rule coefficients
     to evaluate a gradient result."""
-
     _, coeffs, fn, unshifted_coeff, batch_size = data
 
     # individual post-processing of e.g. Hamiltonian grad tapes
     if fn is not None:
         res = fn(res)
 
-    # compute the linear combination of results and coefficients
-    axis = 0
-    if not broadcast:
-        res = qml.math.stack(res)
-    elif (
-        qml.math.get_interface(res[0]) != "torch"
-        and batch_size is not None
-        and not scalar_qfunc_output
-    ):
-        # If the original output is not scalar and broadcasting is used, the second axis
-        # (index 1) needs to be contracted. For Torch, this is not true because the
-        # output of the broadcasted tape is flat due to the behaviour of the Torch device.
-        axis = 1
-    g = qml.math.tensordot(res, qml.math.convert_like(coeffs, res), [[axis], [0]])
+    if not qml.active_return():
 
-    if unshifted_coeff is not None:
-        # add the unshifted term
-        g = g + unshifted_coeff * r0
+        axis = 0
+        if not broadcast:
+            res = qml.math.stack(res)
+        elif (
+            qml.math.get_interface(res[0]) != "torch"
+            and batch_size is not None
+            and not scalar_qfunc_output
+        ):
+            # If the original output is not scalar and broadcasting is used, the second axis
+            # (index 1) needs to be contracted. For Torch, this is not true because the
+            # output of the broadcasted tape is flat due to the behaviour of the Torch device.
+            axis = 1
+
+        # compute the linear combination of results and coefficients
+        g = qml.math.tensordot(res, qml.math.convert_like(coeffs, res), [[axis], [0]])
+
+        g = _unshifted_coeff(g, unshifted_coeff, r0)
+
+    else:
+        multi_measure = len(tape.measurements) > 1
+        if not multi_measure:
+
+            res = qml.math.stack(res)
+
+            if (
+                tape.measurements[0].return_type == qml.measurements.Probability
+                or len(res.shape) > 1
+            ):

```

Comment:
Is the first case included in the second here? If the measurement is `qml.measurements.Probability`, then `res` will have shape `(num_tapes, 2 ** wires)` right?

Reply:
Yes! Thank you, updated.